contracts/CommonHealthCheck.sol

```diff
@@ -0,0 +1,169 @@
+// SPDX-License-Identifier: GPL-3.0
+pragma solidity >=0.6.0 <0.7.0;
+pragma experimental ABIEncoderV2;
+
+struct StrategyParams {
+    uint256 performanceFee;
+    uint256 activation;
+    uint256 debtRatio;
+    uint256 minDebtPerHarvest;
+    uint256 maxDebtPerHarvest;
+    uint256 lastReport;
+    uint256 totalDebt;
+    uint256 totalGain;
+    uint256 totalLoss;
+}
+
+interface VaultAPI {
+    function strategies(address _strategy)
+        external
+        view
+        returns (StrategyParams memory);
+}
+
+interface StrategyAPI {
+    function vault() external view returns (address);
+}
+
+interface CustomHealthCheck {
+    function check(
+        uint256 profit,
+        uint256 loss,
+        uint256 debtPayment,
+        uint256 debtOutstanding,
+        address callerStrategy
+    ) external view returns (bool);
+}
+
+// LEGACY INTERFACES PRE 0.3.2
+struct LegacyStrategyParams {
+    uint256 performanceFee;
+    uint256 activation;
+    uint256 debtRatio;
+    uint256 rateLimit;
+    uint256 lastReport;
+    uint256 totalDebt;
+    uint256 totalGain;
+    uint256 totalLoss;
+}
+
+interface LegacyVaultAPI {
+    function strategies(address _strategy)
+        external
+        view
+        returns (LegacyStrategyParams memory);
+}
+
+contract CommonHealthCheck {
+    // Global Settings for all strategies
+    uint256 constant MAX_BPS = 10_000;
+    uint256 public profitLimitRatio;
+    uint256 public lossLimitRatio;
+
+    address public governance;
+    address public management;
+
+    mapping(address => address) public checks;
+
+    modifier onlyGovernance() {
+        require(msg.sender == governance, "!authorized");
+        _;
+    }
+
+    modifier onlyAuthorized() {
+        require(
+            msg.sender == governance || msg.sender == management,
+            "!authorized"
+        );
+        _;
+    }
+
+    constructor(address _vault) public {
+        governance = msg.sender;
+        management = msg.sender;
+    }
+
+    function setGovernance(address _governance) external onlyGovernance {
+        require(_governance != address(0));
+        governance = _governance;
+    }
+
+    function setManagement(address _management) external onlyGovernance {
+        require(_management != address(0));
+        management = _management;
+    }
+
+    function setCheck(address _strategy, address _check)
+        external
+        onlyAuthorized
+    {
+        checks[_strategy] = _check;
+    }
+
+    function check(
+        uint256 profit,
+        uint256 loss,
+        uint256 debtPayment,
+        uint256 debtOutstanding
+    ) external view returns (bool) {
+        address vault = StrategyAPI(msg.sender).vault();
+        uint256 totalDebt = VaultAPI(vault).strategies(address(this)).totalDebt;
```

Comment:
Something I am not quite about,
Here the `totalDebt` has already been updated by the `adjustPosition` call.

I think that can be an issue but I can't find an example where it's an issue.

Reply:
gotcha, yeah maybe we can record the previous totalDebt and send it as arg in check?

Reply:
yes, that's what I though.

Reply:
the other approach is to add an extra method call at the beginning of harvest on healthcheck

```
// saves state we are interested in prev to harvest side effects
HealthCheck(addr).collectData();
...
// do harvest accounting
..
HealthCheck(addr).check(...);

```